### PR TITLE
storing asciimath and latex for each step (using the node.toTex()) function

### PR DESCRIPTION
**Pull request by aelnaiem**
Does math.js print asciimath? Maybe we don't have to do our own pretty printing? Anyway I think this can be cleaned up, but just putting it up before I call it a day. Pretty straightforward

Part of [this task](https://app.asana.com/0/157236840334946/195847872672803)

### CODE REVIEW

**Comment by evykassirer** [comment]
mathjs does print asciimath, but not for operation nodes with more than 2 children, which is why I had to make my own :(

File: lib/expression/step-solver/Equation.js [hunk 1]
```diff
@@ -17,7 +17,16 @@ class Equation {
   prettyPrint(showPlusMinus=false) {
     let leftSide = prettyPrintNode(this.leftNode, showPlusMinus);
     let rightSide = prettyPrintNode(this.rightNode, showPlusMinus);
-    return leftSide + ' ' + this.comparator + ' ' + rightSide;
+
+    return `${leftSide} ${this.comparator} ${rightSide}`;
```
Comment (evykassirer) [inline lib/expression/step-solver/Equation.js]: love it :D keep forgetting that's a thing

File: lib/expression/step-solver/Equation.js [hunk 2]
```diff
@@ -17,7 +17,16 @@ class Equation {
...
+  },
+
+  // Prints an Equation as LaTeX
+  toTex() {
+    let leftSide = this.leftNode.toTex();
```
Comment (evykassirer) [inline lib/expression/step-solver/Equation.js]: double check that this works for flattened operations with more than 2 children

Comment (evykassirer) [inline lib/expression/step-solver/Equation.js]: (if it doesn't, you could reparse it before printing, so it's not flattened, I guess)

File: lib/expression/step-solver/simplifyExpression.js [hunk 3]
```diff
@@ -74,8 +74,8 @@ function stepThrough(node, debug=false, firstTime=true) {
   // Update the last step, in case we returned something simplified the last
   // time that didn't count as a step.
   if (steps.length > 0) {
```
Comment (evykassirer) [inline lib/expression/step-solver/simplifyExpression.js]: the hasChanged vs isStep change will fix that

File: lib/expression/step-solver/prettyPrint.js [hunk 4]
```diff
@@ -1,71 +1,75 @@
...
+function prettyPrint(node, latex=false, showPlusMinus=false) {
+  let string = prettyPrintDFS(node, false);
+  if (!showPlusMinus) {
+    string = string.replace(/\s*?\+\s*?\-\s*?/g, ' - ');
   }
-  else {
-    return string.replace(/\+ \-/g, '- ');
+
+  if (latex) {
+    string = math.parse(string).toTex();
```
Comment (evykassirer) [inline lib/expression/step-solver/prettyPrint.js]: add a comment for why we need to do this

File: lib/expression/step-solver/prettyPrint.js [hunk 5]
```diff
@@ -1,71 +1,75 @@
...
 // If it's false (the default) 2 + -3 would print as 2 - 3
 // This supports the conversion of subtraction to addition of negative terms,
 // which is needed to flatten operands.
-function prettyPrint(node, showPlusMinus=false) {
-  let string = prettyPrintDFS(node);
-  if (showPlusMinus) {
-    return string;
+function prettyPrint(node, latex=false, showPlusMinus=false) {
+  let string = prettyPrintDFS(node, false);
+  if (!showPlusMinus) {
```
Comment (aelnaiem) [inline lib/expression/step-solver/prettyPrint.js]: Now it does! That's was my original mistake

Comment (evykassirer) [inline lib/expression/step-solver/prettyPrint.js]: ohh cause it's default false. right

can you make some tests? I think there are some for pretty print already.

File: lib/expression/step-solver/prettyPrint.js [hunk 6]
```diff
@@ -14,12 +14,13 @@ const NodeType = require('./NodeType');
...
 // which is needed to flatten operands.
 function prettyPrint(node, latex=false, showPlusMinus=false) {
-  let string = prettyPrintDFS(node, false);
+  let string = prettyPrintDFS(node);
   if (!showPlusMinus) {
     string = string.replace(/\s*?\+\s*?\-\s*?/g, ' - ');
   }
 
   if (latex) {
+    // we parse our string and use the mathjs node TeX printer to get the latex
```
Comment (evykassirer) [inline lib/expression/step-solver/prettyPrint.js]: can there be something about how node might be flattened and then the TeX won't support it?

**Comment by evykassirer** [comment]
sweet just one more thing - can you add some tests? since if this stuff breaks that'd be pretty bad

**Comment by evykassirer** [comment]
👍

**Pull request merged**